test/gtest/ucp/test_ucp_proto.cc

```diff
@@ -343,3 +344,218 @@ UCS_TEST_P(test_perf_node, replace_null)
+{
+    UCS_STRING_BUFFER_ONSTACK(strb, 256);
+    ucp_proto_perf_dump(perf, &strb);
+    return os << ucs_string_buffer_cstr(&strb);
+}
+
+class test_proto_perf : public test_ucp_proto {
+protected:
+    ucp_proto_perf_segment_t *find_lb(size_t start) const
+    {
+        return ucp_proto_perf_find_segment_lb(m_perf, start);
+    }
+
+    void expect_empty_range(size_t start, size_t end)
+    {
+        ucp_proto_perf_segment_t *seg = find_lb(start);
+
+        EXPECT_TRUE((seg == nullptr) ||
+                    (end < ucp_proto_perf_segment_start(seg))) <<
+                    "perf=" << m_perf << " start=" << start << " end=" << end;
+    }
+
+    void expect_perf(size_t start, size_t end,
+                     ucp_proto_perf_factor_id_t factor_id,
+                     ucs_linear_func_t func)
+    {
+        ucp_proto_perf_segment_t *seg = find_lb(start);
+        ASSERT_NE(nullptr, seg);
+        EXPECT_GE(start, ucp_proto_perf_segment_start(seg));
+        EXPECT_LE(end, ucp_proto_perf_segment_end(seg));
+        EXPECT_TRUE(ucs_linear_func_is_equal(
+                func, ucp_proto_perf_segment_func(seg, factor_id), 1e-9));
+    }
+
+    void
+    expect_perf(size_t start, size_t end,
+                const std::initializer_list<
+                        std::pair<ucp_proto_perf_factor_id_t, ucs_linear_func_t>>
+                        factors)
+    {
+        uint64_t zero_perf_factors = UCS_MASK(UCP_PROTO_PERF_FACTOR_LAST);
+        for (auto factor : factors) {
+            expect_perf(start, end, factor.first, factor.second);
+            zero_perf_factors &= ~UCS_BIT(factor.first);
+        }
+
+        unsigned factor_id;
+        ucs_for_each_bit(factor_id, zero_perf_factors) {
+            expect_perf(start, end,
+                        static_cast<ucp_proto_perf_factor_id_t>(factor_id),
+                        UCS_LINEAR_FUNC_ZERO);
+        }
+    }
+
+protected:
+    const ucs_linear_func_t tl_send_func   = {10, 1};
+    const ucs_linear_func_t tl_recv_func   = {20, 2};
+    const ucs_linear_func_t local_cpu_func = {30, 3};
+    ucp_proto_perf_t *m_perf               = nullptr;
+};
+
+UCS_TEST_P(test_proto_perf, from_caps)
+{
+    ucp_proto_caps_t caps;
+
+    caps.min_length = 1000;
+    caps.num_ranges = 2;
+
+    caps.ranges[0].max_length                       = 1999;
+    caps.ranges[0].perf[UCP_PROTO_PERF_TYPE_SINGLE] = tl_send_func;
+    caps.ranges[0].perf[UCP_PROTO_PERF_TYPE_MULTI]  = UCS_LINEAR_FUNC_ZERO;
+    caps.ranges[0].perf[UCP_PROTO_PERF_TYPE_CPU]    = UCS_LINEAR_FUNC_ZERO;
+    caps.ranges[0].node                             = NULL;
+
+    caps.ranges[1].max_length                       = 2999;
+    caps.ranges[1].perf[UCP_PROTO_PERF_TYPE_SINGLE] = UCS_LINEAR_FUNC_ZERO;
+    caps.ranges[1].perf[UCP_PROTO_PERF_TYPE_MULTI]  = UCS_LINEAR_FUNC_ZERO;
+    caps.ranges[1].perf[UCP_PROTO_PERF_TYPE_CPU]    = local_cpu_func;
+    caps.ranges[1].node                             = NULL;
```

Comment:
I understand that patch is huge already, but WDYTA adding perf nodes allocation to check how they splitted/allocated/deleted?

Reply:
it's tested to some extend in the intersection tests, IMO it's better to make more testing in a separate PR